[MCP] Added parameter description.

anushakolan opened the pull request:
## Why make this change?

- Expands stored procedure parameter definitions in the configuration file to support richer metadata, using parameter's attributes like, `name`, `description`, `required`, and `default` values.
- Enables MCP and other tooling to leverage parameter metadata for improved documentation, validation, and user experience.
- Supports backward compatibility by allowing both the legacy dictionary format and the new array-of-objects format for parameters.

## What is this change?

- Changes the `parameters` property in entity config from a dictionary mapping parameter names to values, to an array of parameter objects with explicit fields (`name`, `required`, `default`, `description`).
- Updates the JSON schema to support the new structure and maintain support for the old syntax.
- Refactors the internal model to handle both formats and expose richer metadata.
- Enhances CLI commands (`dab add` and `dab update`) to accept and manage parameter metadata fields: `name`, `required`, `default`, and `description`.
- Adds support for the expanded parameter object in CLI validation, including deprecation notes for the legacy format.
- Includes parameter descriptions in DAB's OpenAPI and GraphQL schema outputs for improved API documentation.
- This change also introduces smarter handling of optional parameters for stored procedures. If a parameter is not required and neither a user value nor a config default is provided, DAB will skip sending that parameter, allowing the database’s internal default to apply. This avoids duplication and drift between config and database defaults. The old behavior is fully supported for backward compatibility.

## How was this tested?

- [x] Manual Testing:
-  Used CLI commands to add, validate and update parameters with metadata.
-  Verified config parsing, CLI commands, and schema generation for both legacy and expanded parameter formats.
- Confirmed correct behavior in OpenAPI and GraphQL outputs.

## Sample Request(s)

**Before (legacy dictionary format):**
```json
{
  "entities": {
    "User": {
      "source": {
        "parameters": {
          "param1": "default-value"
        }
      }
    }
  }
}
```

**After (expanded array format):**
```json
{
  "entities": {
    "User": {
      "source": {
        "parameters": [
          {
            "name": "Id",
            "required": true,
            "default": "default-value",
            "description": "The unique identifier for the user."
          }
        ]
      }
    }
  }
}
```

**Command Line Examples:**

**Before (legacy dictionary format):**

```
dotnet C:\DAB\data-api-builder\src\out\cli\net8.0\Microsoft.DataApiBuilder.dll add MyProc --source "dbo.MyProc" --source.type stored-procedure --source.params "PageSize:50,SortOrder:ASC" --permissions "anonymous:execute" "authenticated:execute" --config C:\DAB\data-api-builder\src\Service\dab-config.json 
```

```
"MyProc": {
      "source": {
        "object": "dbo.MyProc",
        "type": "stored-procedure",
        "parameters": [
          {
            "name": "PageSize",
            "required": false,
            "default": "50"
          },
          {
            "name": "SortOrder",
            "required": false,
            "default": "ASC"
          }
        ]
      },
      "graphql": {
        "enabled": true,
        "operation": "mutation",
        "type": {
          "singular": "MyProc",
          "plural": "MyProcs"
        }
      },
      "rest": {
        "enabled": true,
        "methods": [
          "post"
        ]
      },
      "permissions": [
        {
          "role": "anonymous",
          "actions": [
            {
              "action": "execute"
            }
          ]
        }
      ]
    }
```
 **After (expanded array format):**

Add:
```
dotnet C:\DAB\data-api-builder\src\out\cli\net8.0\Microsoft.DataApiBuilder.dll add entity GetTodosByOwnerAndStatus --source "dbo.GetTodosByOwnerAndStatus" --source.type stored-procedure --parameters.name "OwnerId,Completed" --parameters.description "Owner ID,Completed status" --parameters.required "true,false" --parameters.default "public,0" --permissions "anonymous:execute" --config C:\DAB\data-api-builder\src\Service\dab-config.json
```
Config created:
``` 
"entity": {
      "source": {
        "object": "dbo.GetTodosByOwnerAndStatus",
        "type": "stored-procedure",
        "parameters": [
          {
            "name": "OwnerId",
            "description": "Owner ID",
            "required": true,
            "default": "public"
          },
          {
            "name": "Completed",
            "description": "Completed status",
            "required": false,
            "default": "0"
          }
        ]
      },
      "graphql": {
        "enabled": true,
        "operation": "mutation",
        "type": {
          "singular": "entity",
          "plural": "entities"
        }
      },
      "rest": {
        "enabled": true,
        "methods": [
          "post"
        ]
      },
      "permissions": [
        {
          "role": "anonymous",
          "actions": [
            {
              "action": "execute"
            }
          ]
        }
      ]
    }
```
Validate:
```
dotnet C:\DAB\data-api-builder\src\out\cli\net8.0\Microsoft.DataApiBuilder.dll   validate -c "C:\DAB\data-api-builder\src\Service\dab-config.json" 
```

Validate when both formats are given:
```

PS C:\DAB\data-api-builder\src\Service> dotnet C:\DAB\data-api-builder\src\out\cli\net8.0\Microsoft.DataApiBuilder.dll add MyProc --source "dbo.MyProc" --source.type stored-procedure --source.params "PageSize:50,SortOrder:ASC" --permissions "anonymous:execute" "authenticated:execute" --config C:\DAB\data-api-builder\src\Service\dab-config.json --parameters.name "OwnerId,Completed" --parameters.description "Owner ID,Completed status" --parameters.required "true,false" --parameters.default "public,0" 
Information: Microsoft.DataApiBuilder 1.7.0
Information: User provided config file: C:\DAB\data-api-builder\src\Service\dab-config.json
Loading config file from C:\DAB\data-api-builder\src\Service\dab-config.json.    
Error: Cannot use both --source.params and --parameters.name/description/required/default together. Please use only one format.
Error: Unable to create the source object.
Error: Failed to add a new entity.
Error: Could not add entity: MyProc with source: dbo.MyProc and permissions: anonymous:execute.
```

Update:
```
dotnet C:\DAB\data-api-builder\src\out\cli\net8.0\Microsoft.DataApiBuilder.dll update entity GetTodosByOwnerAndStatus --parameters.name "OwnerId,Completed" --parameters.description "Owner ID,Completed status" --parameters.required "true,false" --parameters.default "private,protected" --config "C:\DAB\data-api-builder\src\Service\dab-config.json" 
```

Config updated:

```
"entity": {
      "source": {
        "object": "dbo.GetTodosByOwnerAndStatus",
        "type": "stored-procedure",
        "parameters": [
          {
            "name": "OwnerId",
            "description": "Owner ID",
            "required": true,
            "default": "private"
          },
          {
            "name": "Completed",
            "description": "Completed status",
            "required": false,
            "default": "protected"
          }
        ]
      },
      "graphql": {
        "enabled": true,
        "operation": "mutation",
        "type": {
          "singular": "entity",
          "plural": "entities"
        }
      },
      "rest": {
        "enabled": true,
        "methods": [
          "post"
        ]
      },
      "permissions": [
        {
          "role": "anonymous",
          "actions": [
            {
              "action": "execute"
            }
          ]
        }
      ]
    }
```
-----------------------------------------------

Create a stored procedure:

```
CREATE PROCEDURE dbo.InsertTodo
    @id UNIQUEIDENTIFIER,
    @title NVARCHAR(1000),
    @completed BIT = 0,
    @owner_id VARCHAR(128),
    @position INT = NULL
AS
BEGIN
    INSERT INTO dbo.todos (id, title, completed, owner_id, position)
    VALUES (@id, @title, @completed, @owner_id, @position);

    SELECT id, title, completed, owner_id, position
    FROM dbo.todos
    WHERE id = @id;
END
GO
```

1. Without providing default parameter values
```
mutation {
  executeInsertTodo(
    id: "00000000-0000-0000-0000-000000000013",
    title: "Test DAB GraphQL",
    owner_id: "Anusha Kolan"
    # completed and position are optional
  ) {
    id
    title
    completed
    owner_id
    position
  }
}
```

Uses DB default values
```
{
  "data": {
    "executeInsertTodo": [
      {
        "id": "00000000-0000-0000-0000-000000000013",
        "title": "Test DAB GraphQL",
        "completed": false,
        "owner_id": "Anusha Kolan",
        "position": null
      }
    ]
  }
}
```

2. With providing default parameter values

```
mutation {
  executeInsertTodo(
    id: "00000000-0000-0000-0000-000000000014",
    title: "Test with completed",
    owner_id: "Anusha Kolan",
    completed: true,
    position: 5
  ) {
      id
      title
      completed
      owner_id
      position
  }
}
```

Uses the default values provided

```
{
  "data": {
    "executeInsertTodo": [
      {
        "id": "00000000-0000-0000-0000-000000000014",
        "title": "Test with completed",
        "completed": true,
        "owner_id": "Anusha Kolan",
        "position": 5
      }
    ]
  }
}
```
-----------------------------------------------------------------------------------

Config:

```
"InsertTodo": {
      "source": {
        "object": "dbo.InsertTodo",
        "type": "stored-procedure",
        "parameters": [
          {
            "name": "id",
            "description": "The id for the Todo.",
            "required": true
          },
          {
            "name": "title",
            "description": "The title for the Todo.",
            "required": true
          },
          {
            "name": "completed",
            "description": "The state of the Todo.",
            "required": false
          },
          {
            "name": "owner_id",
            "required": true
          },
          {
            "name": "position",
            "required": false
          }
        ]
      },
      "graphql": {
        "enabled": true,
        "operation": "mutation",
        "type": {
          "singular": "InsertTodo",
          "plural": "InsertTodos"
        }
      },
      "rest": {
        "enabled": true,
        "methods": [
          "post"
        ]
      },
      "permissions": [
        {
          "role": "anonymous",
          "actions": [
            {
              "action": "execute"
            }
          ]
        }
      ]
    }
```

GraphQL Schema Documentation

```
"Execute Stored-Procedure InsertTodo and get results from the database"
  executeInsertTodo(
    "The state of the Todo."
    completed: Boolean
    "The id for the Todo."
    id: UUID
    "parameters for InsertTodo stored-procedure"
    owner_id: String
    "parameters for InsertTodo stored-procedure"
    position: Int
    "The title for the Todo."
    title: String
  ): [InsertTodo!]! @cost(weight: "10")
```

OpenAPI Documentation

```
"InsertTodo_sp_request": {
        "required": [
          "id",
          "owner_id",
          "title"
        ],
        "type": "object",
        "properties": {
          "completed": {
            "type": "boolean",
            "description": "The state of the Todo."
          },
          "id": {
            "type": "string",
            "description": "The id for the Todo."
          },
          "owner_id": {
            "type": "string"
          },
          "position": {
            "type": "number"
          },
          "title": {
            "type": "string",
            "description": "The title for the Todo."
          }
        }
      },
```

souvikghosh04 reviewed (approved):
Approving this. Please address the open comments and test them.

RubenCerna2079 commented on src/Cli.Tests/UtilsTests.cs:
```diff
@@ -137,13 +137,13 @@ public void TestConfigSelectionBasedOnCliPrecedence(
     {
         IEnumerable<string>? parametersList = new string[] { "param1:123", "param2:-243", "param3:220.12", "param4:True", "param5:dab" };
-        Assert.IsTrue(TryParseSourceParameterDictionary(parametersList, out Dictionary<string, object>? sourceParameters));
+        Assert.IsTrue(TryParseSourceParameterDictionary(parametersList, out List<Azure.DataApiBuilder.Config.ObjectModel.ParameterMetadata>? sourceParameters));
```
Is it possible to test the legacy way as well?

RubenCerna2079 commented on src/Cli.Tests/Snapshots/AddEntityTests.AddNewEntityWhenEntitiesWithSourceAsStoredProcedure.verified.txt:
```diff
@@ -28,11 +28,23 @@
-            param2: hello,
-            param3: true
-          }
+          Parameters: [
```
Just saw later in the code that we can only use one at a time, in that case I think we should create some tests for the legacy option

anushakolan replied:
Considering the time constraints, I will have a follow up task for all the extra testing.

RubenCerna2079 commented on src/Core/Services/MetadataProviders/SqlMetadataProvider.cs:
```diff
@@ -1169,25 +1174,6 @@ private async Task PopulateResultSetDefinitionsForStoredProcedureAsync(
-        /// <param name="paramName">Common prefix of param names.</param>
-        /// <param name="paramValues">Values of the param.</param>
-        /// <returns></returns>
-        private static Dictionary<string, object> GetQueryParams(
```
Thanks for removing this unused method :)

Aniruddh25 commented on schemas/dab.draft.schema.json:
```diff
@@ -755,23 +755,35 @@
+                              "name": { "type": "string", "description": "Parameter name" },
+                              "required": { "type": "boolean", "description": "Is parameter required" },
+                              "default": { "type": ["string", "number", "boolean", "null"], "description": "Default value" },
+                              "description": { "type": "string", "description": "Parameter description" }
```
given the format of accepting description for multiple parameters is a `,` - comma separated string, we should mention here that the description of individual parameters shouldn't contain `,`.

Aniruddh25 commented on src/Config/Converters/EntitySourceConverterFactory.cs:
```diff
@@ -57,75 +57,62 @@ public EntitySourceConverter(bool replaceEnvVar)
+                    ?? throw new JsonException("Failed to deserialize EntitySource from modified JSON.");
+
+                // Use the with expression to set the correct Parameters
+                return entitySource with { Parameters = paramList };
```
Isnt line 89 which sets `modObj["parameters"] = paramList;` already serving the purpose of filling up the entitySource with the new format of params? 

Why is the `with` expression needed here?

Aniruddh25 commented on src/Config/Converters/EntitySourceConverterFactory.cs:
```diff
@@ -57,75 +57,62 @@ public EntitySourceConverter(bool replaceEnvVar)
+                // If already in new format, deserialize as usual
+                EntitySource? source = JsonSerializer.Deserialize<EntitySource>(rawJson, innerOptions);
+
+                if (source?.Parameters is not null)
```
Wouldnt deserialization of rawJson here fill up the source.Parameters with the correct list? What is the purpose of this `with` expression on line 110?

Aniruddh25 commented on src/Config/ObjectModel/ParameterMetadata.cs:
```diff
@@ -0,0 +1,13 @@
+    /// </summary>
+    public class ParameterMetadata
+    {
+        public required string Name { get; set; }
```
Dont we need to provide the Json property to these member properties? like `name` and `description`.

Aniruddh25 commented on src/Cli/ConfigGenerator.cs:
```diff
@@ -1842,49 +1879,155 @@ private static bool TryGetUpdatedSourceObjectWithOptions(
+
+                List<ParameterMetadata> mergedParams = new();
+
+                if (parameters != null)
```
wouldnt `parameters` be not null here, given that it is inside `hasNewParams`

anushakolan replied:
Yes that is right, but I had to add the if condition to fix this error in the for loop, 
```
Dereference of a possibly null reference.[CS8602](https://msdn.microsoft.com/query/roslyn.query?appId=roslyn&k=k(CS8602))
(local variable) List<ParameterMetadata>? parameters
'parameters' may be null here.
```

Aniruddh25 commented on src/Core/Services/MetadataProviders/SqlMetadataProvider.cs:
```diff
@@ -469,8 +470,12 @@ protected virtual async Task FillSchemaForStoredProcedureAsync(
                     {
-                        parameterDefinition.HasConfigDefault = true;
-                        parameterDefinition.ConfigDefaultValue = configParamValue?.ToString();
+                        // Map all metadata from config
```
Eventually, we should be augmenting this metadata in SQL. So, that config has less information and we expose that schema info from the db directly as a tool to the MCP client.

Aniruddh25 reviewed (approved):
Left a few comments for optimizing the code to remove some not required checks. Approved, assuming the comments will be resolved.

anushakolan commented on src/Cli/Commands/EntityOptions.cs:
```diff
@@ -91,5 +100,17 @@ public EntityOptions(
+        public IEnumerable<string>? ParametersRequired { get; }
+
+        [Option("parameters.default", Required = false, Separator = ',', HelpText = "Comma-separated list of parameter default values for stored procedure.")]
+        public IEnumerable<string>? ParametersDefault { get; }
```
Done

anushakolan commented on src/Service.Tests/OpenApiDocumentor/ParameterValidationTests.cs:
```diff
@@ -144,10 +144,23 @@ public async Task TestInputParametersForStoredProcedures()
+                Required = false,
+                Default = "Test Title"
+            }
+        };
```
I just used them similar to previous values

anushakolan commented:
> Left a few comments for optimizing the code to remove some not required checks. Approved, assuming the comments will be resolved.

I have addressed your comments Ani.

anushakolan commented:
/azp run

RubenCerna2079 commented on schemas/dab.draft.schema.json:
```diff
@@ -755,23 +755,35 @@
+                              "required": { "type": "boolean", "description": "Is parameter required" },
+                              "default": { "type": ["string", "number", "boolean", "null"], "description": "Default value" },
+                                "description": { "type": "string", "description": "Parameter description. Since descriptions for multiple parameters are provided as a comma-separated string, individual parameter descriptions must not contain a comma (',')." }
+                            }
```
nit:
```suggestion
                              "default": { "type": ["string", "number", "boolean", "null"], "description": "Default value" },
                              "description": { "type": "string", "description": "Parameter description. Since descriptions for multiple parameters are provided as a comma-separated string, individual parameter descriptions must not contain a comma (',')." }
                            }
```

RubenCerna2079 commented on src/Config/Converters/EntitySourceConverterFactory.cs:
```diff
@@ -68,7 +68,7 @@ public EntitySourceConverter(bool replaceEnvVar)
                 foreach (JsonProperty prop in parametersElement.EnumerateObject())
                 {
                     object? value = prop.Value.Deserialize<object>(innerOptions);
```
I don't see any place where we use this `value` object, so it should be fine to delete it

anushakolan commented on src/Config/Converters/EntitySourceConverterFactory.cs:
```diff
@@ -57,17 +57,62 @@ public EntitySourceConverter(bool replaceEnvVar)
+                List<ParameterMetadata> paramList = [];
+                foreach (JsonProperty prop in parametersElement.EnumerateObject())
+                {
+                    object? value = prop.Value.Deserialize<object>(innerOptions);
```
```suggestion   
```
